task/source-build/0.1/source-build.yaml

```diff
@@ -0,0 +1,472 @@
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: source-build
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: "appstudio, hacbs"
+spec:
+  description: Source image build.
+  params:
+    - name: BINARY_IMAGE
+      description: The corresponding binary image based on which to generate source image.
+      type: string
+    - name: BASE_IMAGES
+      description: >-
+        Base images used to build the binary image. Each image per line in the same order of FROM
+        instructions specified in a multistage Dockerfile. Default to an empty string, which means
+        to skip handling a base image.
+      type: string
+      default: ""
+  results:
+    - name: BUILD_RESULT
+      description: Build result.
+    - name: SOURCE_IMAGE_URL
+      description: The source image url.
+  workspaces:
+  - name: workspace
+    description: Workspace where task works inside. Source code is included in this workspace and any artifacts generated by this task are also stored here.
+  volumes:
+    - name: source-build-work-place
+      emptyDir:
+        medium: Memory
+  steps:
+    - name: build
+      image: quay.io/redhat-appstudio/build-definitions-source-image-build-utils:latest
+      resources:
+        limits:
+          memory: 4Gi
+          cpu: 1
+        requests:
+          memory: 512Mi
+          cpu: 250m
+      workingDir: "/var/source-build"
+      securityContext:
+        runAsUser: 0
+        capabilities:
+          add:
+            - SETFCAP
+      volumeMounts:
+        - name: source-build-work-place
+          mountPath: /var/source-build
+      args:
+        - "--output-binary-image"
+        - "$(params.BINARY_IMAGE)"
+        - "--workspace"
+        - "/var/source-build"
+        - "--source-dir"
+        - "$(workspaces.workspace.path)/source"
+        - "--base-images"
+        - "$(params.BASE_IMAGES)"
+        - "--write-result-to"
+        - "$(results.BUILD_RESULT.path)"
+        - "--cachi2-artifacts-dir"
+        - "$(workspaces.workspace.path)/cachi2"
+      script: |
+        #!/usr/bin/python3.11
+
+        import argparse
+        import functools
+        import json
+        import itertools
+        import os
+        import shutil
+        import logging
+        import stat
+        import sys
+        import tarfile
+        import tempfile
+        import urllib.parse
+        from dataclasses import dataclass, field
+        from subprocess import run, CalledProcessError
+        from typing import TypedDict, NotRequired, Literal, Final
+
+        """
+        Requires: git, skopeo, tar, BuildSourceImage
+        """
+
+        logging.basicConfig(level=logging.DEBUG, format="%(asctime)s:%(name)s:%(levelname)s:%(message)s")
```

Comment:
will it be good to change logging level to `INFO` when testing is complete?

Reply:
I would keep it in DEBUG so that we can get detailed information in case some issue happens in prod.